crates/typst-macros/src/elem.rs

```diff
@@ -480,88 +472,904 @@ fn create_param_info(field: &Field) -> TokenStream {
-                    element.#push_ident(#value);
+                    #enum_ident::#ident => {
+                        ::typst::diag::bail!(#unsettable_field);
+                    }
                 }
             }
         });
 
-    quote! {
-        impl ::typst::model::Construct for #ident {
-            fn construct(
-                vm: &mut ::typst::eval::Vm,
-                args: &mut ::typst::eval::Args,
-            ) -> ::typst::diag::SourceResult<::typst::model::Content> {
-                let mut element = Self(::typst::model::Content::new(
-                    <Self as ::typst::model::NativeElement>::elem()
-                ));
-                #(#handlers)*
-                Ok(element.0)
-            }
+    // Statistically compute whether we need preparation or not.
+    let needs_preparation = element
+        .without_capability("Unlabellable", || {
+            element
+                .capabilities
+                .iter()
+                .any(|capability| capability == "Locatable" || capability == "Synthesize")
+                .then(|| quote! { !self.prepared })
+                .unwrap_or_else(|| quote! { self.label().is_some() && !self.prepared })
+        })
+        .unwrap_or_else(|| {
+            assert!(element.capabilities.iter().all(|capability| capability
+                != "Locatable"
+                && capability != "Synthesize"));
+            quote! { false }
+        });
+
+    // Creation of the fields dictionary for inherent fields.
+    let field_dict = element.inherent_fields().clone().map(|field| {
+        let name = &field.name;
+        let field_ident = &field.ident;
+
+        quote! {
+            fields.insert(
+                EcoString::inline(#name).into(),
+                ::typst::eval::IntoValue::into_value(self.#field_ident.clone())
+            );
         }
-    }
-}
+    });
 
-/// Create the element's `Set` implementation.
-fn create_set_impl(element: &Elem) -> TokenStream {
-    let ident = &element.ident;
-    let handlers = element
-        .fields
-        .iter()
-        .filter(|field| {
-            !field.external
-                && !field.synthesized
-                && field.settable()
-                && (!field.internal || field.parse.is_some())
-        })
+    // Creation of the fields dictionary for optional fields.
+    let field_opt_dict = element
+        .visible_fields()
+        .filter(|field| !field.inherent())
+        .clone()
         .map(|field| {
-            let set_ident = &field.set_ident;
-            let (prefix, value) = create_field_parser(field);
+            let name = &field.name;
+            let field_ident = &field.ident;
+
             quote! {
-                #prefix
-                if let Some(value) = #value {
-                    styles.set(Self::#set_ident(value));
+                if let Some(value) = &self.#field_ident {
+                    fields.insert(
+                        EcoString::inline(#name).into(),
```

Comment:
EcoString::inline could panic if the field name is longer than 15 bytes. Just .into() is correct here. The `inline` function is only intended for const-only contexts.